crates/duty-tracker/src/contract_manager.rs

```diff
@@ -405,35 +422,45 @@ impl ContractManager {
-                                    // in case an error occurs, we will just nag again
-                                    // so no need to break out of the event loop
+                    Ok(GossipEvent::ReceivedMessage(raw_msg)) = gossip_handle.next_event() => {
+                        match GossipsubMsg::from_bytes(&raw_msg) {
+                            Ok(msg) => {
+                                match ctx.process_p2p_message(msg.clone()).await {
+                                    Ok(msg_duties) if !msg_duties.is_empty() => {
+                                        duties.extend(msg_duties);
+                                    },
+                                    Ok(_) => {},
+                                    Err(e) => {
+                                        error!(?msg, %e, "failed to process p2p msg");
+                                        // in case an error occurs, we will just nag again
+                                        // so no need to break out of the event loop
+                                    }
                                 }
                             }
-                        },
-                        Ok(Event::ReceivedRequest(req)) => {
-                            match ctx.process_p2p_request(req.clone()).await {
-                                Ok(p2p_duties) => duties.extend(p2p_duties),
-                                Err(e) => {
-                                    error!(?req, %e, "failed to process p2p request");
-                                    // in case an error occurs, the requester will just nag again
-                                    // so no need to break out of the event loop
-                                },
+                            Err(e) => {
+                                error!(%e, "failed to decode p2p msg");
+                            }
+                        }
+                    },
+
+                    Some(ReqRespEvent::ReceivedRequest(raw_req, sender)) = req_resp_handle.next_event() => {
+                        match Message::decode(raw_req.as_slice()).and_then(GetMessageRequest::from_msg) {
+                            Ok(req) => {
+                                match ctx.process_p2p_request(req.clone()).await {
+                                    Ok(p2p_duties) => duties.extend(p2p_duties),
+                                    Err(e) => {
+                                        error!(?req, ?sender, %e, "failed to process p2p request");
+                                        // in case an error occurs, the requester will just nag again
+                                        // so no need to break out of the event loop
+                                    },
+                                }
+                            }
+                            Err(e) => {
+                                error!(%e, ?sender, "failed to decode p2p request");
                             }
```

Comment:
Now you have a `oneshot` channel in `strata-p2p`'s `ReqRespEvent`.
See the `Sender` in https://github.com/alpenlabs/strata-p2p/blob/7197bde9bf70680d9c11729aa4a43f5f91eaa091/crates/p2p/src/events.rs#L17-L26

```rs
/// Events emitted from the request/response protocol.
#[cfg(feature = "request-response")]
#[derive(Debug)]
pub enum ReqRespEvent {
    /// Received a request from other peer.
    ReceivedRequest(Vec<u8>, Sender<Vec<u8>>),

    /// Received a response from other peer.
    ReceivedResponse(Vec<u8>),
}
```

Hence you can forward the response channel in the P2P duty and you don't need to regossip ReqResp stuff and you can answer straight on a 1-on-1 basis to the requester.